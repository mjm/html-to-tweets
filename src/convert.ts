import * as htmlparser from "htmlparser2"
import { URL } from "url"

export interface ConvertOptions {
  baseURL?: string
}

export function convert(
  html: string,
  options: ConvertOptions = {}
): ConvertResult {
  const handler = new TweetHandler(options)

  const parser = new htmlparser.Parser(handler as any, { decodeEntities: true })
  parser.write(html)
  parser.end()

  return handler.result
}

export interface ConvertResult {
  body: string
  mediaURLs: string[]
}

class TweetHandler {
  options: ConvertOptions
  text: string = ""
  mediaURLs: string[] = []
  urls: string[] = []
  linkedTwitterUser: string | null = null
  embeddedTweet: { url?: string } | null = null

  constructor(options: ConvertOptions) {
    this.options = options
  }

  onopentag(name: string, attrs: { [key: string]: string }) {
    switch (name) {
      case "a":
        if (attrs.href) {
          if (this.embeddedTweet) {
            if (!attrs.href.startsWith("https://twitter.com")) {
              return
            }

            this.embeddedTweet.url = attrs.href.replace(/(.*)\?.*$/, "$1")
            return
          }
          const twitterUser = getTwitterUsername(attrs.href)
          if (twitterUser) {
            this.linkedTwitterUser = twitterUser
          } else {
            this.addURL(attrs.href)
          }
        }
        break
      case "blockquote":
        if (attrs.class === "twitter-tweet") {
          this.embeddedTweet = {}
        } else {
          this.text += "“"
        }
        break
      case "img":
        if (attrs.src) {
          this.addMediaURL(attrs.src)
        }
        break
      case "iframe":
        if (attrs.src) {
          const youtubeURL = getYoutubeLink(attrs.src)
          if (youtubeURL) {
            this.addURL(youtubeURL)
          }
        }
        break
    }
  }

  ontext(text: string) {
    if (this.linkedTwitterUser || this.embeddedTweet) {
      return
    }

    // collapse all whitespace to a single space
    this.text += text.replace(/\s+/g, " ")
  }

  onclosetag(name: string) {
    switch (name) {
      case "p":
        this.text += "\n\n"
        break
      case "br":
        this.text += "\n"
        break
      case "a":
        if (this.linkedTwitterUser) {
          this.text += `@${this.linkedTwitterUser}`
          this.linkedTwitterUser = null
        }
        break
      case "blockquote":
        if (this.embeddedTweet) {
          if (this.embeddedTweet.url) {
            this.addURL(this.embeddedTweet.url)
          }
          this.embeddedTweet = null
        } else {
          this.text += "”"
        }
        break
    }
  }

  onend() {
    this.text = this.text.trim()
    this.text += ` ${this.urls.join(" ")}`
    this.text = this.text
      .split("\n")
      .map(l => l.trim())
      .join("\n")
  }

  get result(): ConvertResult {
    return {
      body: this.text,
      mediaURLs: this.mediaURLs,
    }
  }

  private addURL(url: string) {
    this.urls.push(this.fixURL(url))
  }

  private addMediaURL(url: string) {
    this.mediaURLs.push(this.fixURL(url))
  }

  private fixURL(url: string): string {
    return new URL(url, this.options.baseURL).toString()
  }
}

const twitterUserRegex = new RegExp(
  "^https?://(?:www.)?twitter.com/([A-Za-z0-9_]{1,15})/?$"
)

function getTwitterUsername(url: string): null | string {
  const match = url.match(twitterUserRegex)
  if (!match) {
    return null
  }

  return match[1]
}

const youtubeRegex = new RegExp(
  "^https://www.youtube.com/embed/([A-Za-z0-9_]+)"
)

function getYoutubeLink(url: string): null | string {
  const match = url.match(youtubeRegex)
  if (!match) {
    return null
  }

  const videoID = match[1]
  return `https://www.youtube.com/watch?v=${videoID}`
}
