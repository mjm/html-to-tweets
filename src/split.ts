import { parseTweet, ParsedTweet } from "twitter-text"
import { Tweet } from "./"
import { ConvertResult } from "./convert"

export function split(input: ConvertResult): Tweet[] {
  const [first, ...rest] = numberTweets(splitTweets(input.body))
  return [
    { action: "tweet", body: first, mediaURLs: input.mediaURLs },
    ...rest.map((t): Tweet => ({ action: "tweet", body: t, mediaURLs: [] })),
  ]
}

function splitTweets(text: string, tweets: string[] = []): string[] {
  const [first, rest] = splitOnce(text)
  tweets.push(first)

  if (rest) {
    return splitTweets(rest, tweets)
  }

  return tweets
}

function splitOnce(text: string): [string, string?] {
  const first = shorten(text)

  const second = text.substring(first.length).trim()
  if (second) {
    return [first, second]
  }

  return [first]
}

function shorten(text: string): string {
  let results = parseTweet(text)
  if (!results.valid) {
    // drop a word to make room for numbering
    text = dropWord(truncate(text, results))

    // keep dropping words until it fits
    while ((results = parseTweet(`${text} (XX/XX)`)) && !results.valid) {
      text = dropWord(text)
    }
  }

  return text
}

function truncate(
  text: string,
  { validRangeStart, validRangeEnd }: ParsedTweet
): string {
  return text.substring(validRangeStart, validRangeEnd)
}

function dropWord(text: string): string {
  return text.replace(/\s+\S*$/, "")
}

function numberTweets(tweets: string[]): string[] {
  const total = tweets.length

  if (total <= 1) {
    return tweets
  }

  return tweets.map((t, i) => `${t} (${i + 1}/${total})`)
}
