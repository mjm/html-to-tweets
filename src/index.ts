import { convert } from "./convert"
import { split } from "./split"

interface TranslateInput {
  title: string | null
  url: string
  html: string
}

export type TwitterAction = Tweet | Retweet

export interface Tweet {
  action: "tweet"
  body: string
  mediaURLs: string[]
}

export interface Retweet {
  action: "retweet"
  tweetID: string
}

export function translate(input: TranslateInput): TwitterAction[] {
  if (input.title) {
    return [
      {
        action: "tweet",
        body: `${input.title} ${input.url}`,
        mediaURLs: [],
      },
    ]
  }

  const result = convert(input.html, { baseURL: input.url })
  const tweets = split(result)
  if (tweets.length === 1) {
    const retweetID = extractRetweetId(tweets[0])
    if (retweetID) {
      return [
        {
          action: "retweet",
          tweetID: retweetID,
        },
      ]
    }
  }

  return tweets
}

const tweetRegex = new RegExp(
  "^https?://(?:www.)?twitter.com/[A-Za-z0-9_]+/status/([0-9]+)/?$"
)

function extractRetweetId(tweet: Tweet): string | null {
  if (tweet.mediaURLs.length) {
    // if the tweet has media, it's not a retweet
    return null
  }

  const match = tweet.body.match(tweetRegex)
  if (!match) {
    return null
  }

  return match[1]
}
