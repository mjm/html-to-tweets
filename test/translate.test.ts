import { translate } from "../src"

test("translates titled post", () => {
  const input = {
    title: "Welcome to Microblogging",
    url: "https://example.com/abc/",
    html: "Who cares?",
  }
  expect(translate(input)).toMatchSnapshot()
})

test("translates short microblog post", () => {
  const input = {
    title: "",
    url: "https://example.com/abc/",
    html: "This is a simple <em>tweet</em>.",
  }
  expect(translate(input)).toMatchSnapshot()
})

test("translates a microblog post with images", () => {
  const input = {
    title: "",
    url: "https://example.com/abc/",
    html:
      '<p>This is a tweet with media.</p><figure><img src="/foo.jpg"></figure>',
  }
  expect(translate(input)).toMatchSnapshot()
})

test("translates a long microblog post", () => {
  const input = {
    title: "",
    url: "https://example.com/abc/",
    html: `
    <p>Spicy jalapeno meatloaf shoulder sirloin, cow frankfurter tail pork belly
    leberkas brisket chuck pancetta bresaola. Shankle chuck rump kielbasa burgdoggen.
    Filet mignon turkey doner t-bone bresaola, ribeye biltong. Shank biltong kevin
    pork pancetta frankfurter, turkey tongue pork loin doner turducken drumstick
    jerky boudin t-bone.</p>

    <p>Pancetta shank boudin, tail corned beef tenderloin kevin buffalo sausage
    tri-tip. T-bone landjaeger porchetta corned beef. Spare ribs strip steak bresaola
    ham pancetta prosciutto. T-bone turkey fatback, turducken pork pig cow jerky
    short loin ball tip. Boudin ball tip brisket, shankle chuck ham frankfurter
    hamburger buffalo landjaeger.</p>
    `,
  }

  expect(translate(input)).toMatchSnapshot()
})

test("translates a retweet with a comment", () => {
  const input = {
    title: "",
    url: "https://example.com/abc",
    html: `
    <p>And I&#8217;m sitting here with two level 117 characters.</p>
    <blockquote class="twitter-tweet" data-width="550" data-dnt="true">
    <p lang="en" dir="ltr">Congratulations <a href="https://twitter.com/Methodgg?ref_src=twsrc%5Etfw">@Methodgg</a> on the world first clear of Mythic Uldir!!! <a href="https://t.co/iUCOwDbPYw">pic.twitter.com/iUCOwDbPYw</a></p>
    <p>&mdash; World of Warcraft (@Warcraft) <a href="https://twitter.com/Warcraft/status/1042485327267422208?ref_src=twsrc%5Etfw">September 19, 2018</a></p></blockquote>
    <p><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script></p>
    `,
  }

  expect(translate(input)).toMatchSnapshot()
})

test("translates a retweet without comment", () => {
  const input = {
    title: "",
    url: "https://example.com/abc",
    html: `
    <blockquote class="twitter-tweet" data-width="550" data-dnt="true">
    <p lang="en" dir="ltr">Congratulations <a href="https://twitter.com/Methodgg?ref_src=twsrc%5Etfw">@Methodgg</a> on the world first clear of Mythic Uldir!!! <a href="https://t.co/iUCOwDbPYw">pic.twitter.com/iUCOwDbPYw</a></p>
    <p>&mdash; World of Warcraft (@Warcraft) <a href="https://twitter.com/Warcraft/status/1042485327267422208?ref_src=twsrc%5Etfw">September 19, 2018</a></p></blockquote>
    <p><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script></p>
    `,
  }

  expect(translate(input)).toMatchSnapshot()
})
