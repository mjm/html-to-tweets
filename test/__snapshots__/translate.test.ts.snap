// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translates a long microblog post 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "Spicy jalapeno meatloaf shoulder sirloin, cow frankfurter tail pork belly leberkas brisket chuck pancetta bresaola. Shankle chuck rump kielbasa burgdoggen. Filet mignon turkey doner t-bone bresaola, ribeye biltong. Shank biltong kevin pork pancetta frankfurter, turkey (1/3)",
    "mediaURLs": Array [],
  },
  Object {
    "action": "tweet",
    "body": "tongue pork loin doner turducken drumstick jerky boudin t-bone.

Pancetta shank boudin, tail corned beef tenderloin kevin buffalo sausage tri-tip. T-bone landjaeger porchetta corned beef. Spare ribs strip steak bresaola ham pancetta prosciutto. T-bone turkey fatback, (2/3)",
    "mediaURLs": Array [],
  },
  Object {
    "action": "tweet",
    "body": "turducken pork pig cow jerky short loin ball tip. Boudin ball tip brisket, shankle chuck ham frankfurter hamburger buffalo landjaeger. (3/3)",
    "mediaURLs": Array [],
  },
]
`;

exports[`translates a microblog post with images 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "This is a tweet with media.",
    "mediaURLs": Array [
      "https://example.com/foo.jpg",
    ],
  },
]
`;

exports[`translates a retweet with a comment 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "And Iâ€™m sitting here with two level 117 characters. https://twitter.com/Warcraft/status/1042485327267422208",
    "mediaURLs": Array [],
  },
]
`;

exports[`translates a retweet without comment 1`] = `
Array [
  Object {
    "action": "retweet",
    "tweetID": "1042485327267422208",
  },
]
`;

exports[`translates short microblog post 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "This is a simple tweet.",
    "mediaURLs": Array [],
  },
]
`;

exports[`translates titled post 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "Welcome to Microblogging https://example.com/abc/",
    "mediaURLs": Array [],
  },
]
`;
